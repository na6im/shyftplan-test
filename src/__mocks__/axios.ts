import axios from 'axios';

const mockAxios = jest.genMockFromModule<typeof axios>('axios');
mockAxios.create = jest.fn(() => mockAxios) as any;

export default mockAxios;
