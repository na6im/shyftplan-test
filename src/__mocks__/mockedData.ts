export const eventsData = {
  items: [
    {
      id: 1,
      position: { name: 'Books', color: 'lime', id: 83 },
      startsAt: '2020-08-01T06:00:00.000Z',
      endsAt: '2020-08-01T11:00:00.000Z',
    },
    {
      id: 48,
      position: { name: 'Health', color: 'yellow', id: 72 },
      startsAt: '2020-09-10T06:00:00.000Z',
      endsAt: '2020-09-10T14:00:00.000Z',
    },
    {
      id: 61,
      position: { name: 'Sports', color: 'lime', id: 233 },
      startsAt: '2020-12-24T10:00:00.000Z',
      endsAt: '2020-12-24T18:00:00.000Z',
    },
    {
      id: 95,
      position: { name: 'Sports', color: 'lime', id: 45 },
      startsAt: '2020-06-09T09:00:00.000Z',
      endsAt: '2020-06-09T13:00:00.000Z',
    },
    {
      id: 106,
      position: { name: 'Books', color: 'lime', id: 83 },
      startsAt: '2020-04-22T07:00:00.000Z',
      endsAt: '2020-04-22T13:00:00.000Z',
    },
    {
      id: 134,
      position: { name: 'Health', color: 'yellow', id: 72 },
      startsAt: '2020-01-24T10:00:00.000Z',
      endsAt: '2020-01-24T15:00:00.000Z',
    },
    {
      id: 166,
      position: { name: 'Sports', color: 'lime', id: 45 },
      startsAt: '2020-10-06T07:00:00.000Z',
      endsAt: '2020-10-06T13:00:00.000Z',
    },
    {
      id: 166,
      position: { name: 'Sports', color: 'lime', id: 233 },
      startsAt: '2020-07-14T10:00:00.000Z',
      endsAt: '2020-07-14T14:00:00.000Z',
    },
    {
      id: 195,
      position: { name: 'Sports', color: 'lime', id: 45 },
      startsAt: '2020-11-10T07:00:00.000Z',
      endsAt: '2020-11-10T13:00:00.000Z',
    },
    {
      id: 239,
      position: { name: 'Sports', color: 'lime', id: 45 },
      startsAt: '2020-02-05T06:00:00.000Z',
      endsAt: '2020-02-05T13:00:00.000Z',
    },
  ],
  pagination: { offset: 0, limit: 10, count: 50 },
};

export const eventsDataNoMore = {
  items: [
    {
      id: 1,
      position: { name: 'Books', color: 'lime', id: 83 },
      startsAt: '2020-08-01T06:00:00.000Z',
      endsAt: '2020-08-01T11:00:00.000Z',
    },
    {
      id: 48,
      position: { name: 'Health', color: 'yellow', id: 72 },
      startsAt: '2020-09-10T06:00:00.000Z',
      endsAt: '2020-09-10T14:00:00.000Z',
    },
    {
      id: 61,
      position: { name: 'Sports', color: 'lime', id: 233 },
      startsAt: '2020-12-24T10:00:00.000Z',
      endsAt: '2020-12-24T18:00:00.000Z',
    },
    {
      id: 95,
      position: { name: 'Sports', color: 'lime', id: 45 },
      startsAt: '2020-06-09T09:00:00.000Z',
      endsAt: '2020-06-09T13:00:00.000Z',
    },
    {
      id: 106,
      position: { name: 'Books', color: 'lime', id: 83 },
      startsAt: '2020-04-22T07:00:00.000Z',
      endsAt: '2020-04-22T13:00:00.000Z',
    },
    {
      id: 134,
      position: { name: 'Health', color: 'yellow', id: 72 },
      startsAt: '2020-01-24T10:00:00.000Z',
      endsAt: '2020-01-24T15:00:00.000Z',
    },
    {
      id: 166,
      position: { name: 'Sports', color: 'lime', id: 45 },
      startsAt: '2020-10-06T07:00:00.000Z',
      endsAt: '2020-10-06T13:00:00.000Z',
    },
    {
      id: 166,
      position: { name: 'Sports', color: 'lime', id: 233 },
      startsAt: '2020-07-14T10:00:00.000Z',
      endsAt: '2020-07-14T14:00:00.000Z',
    },
    {
      id: 195,
      position: { name: 'Sports', color: 'lime', id: 45 },
      startsAt: '2020-11-10T07:00:00.000Z',
      endsAt: '2020-11-10T13:00:00.000Z',
    },
    {
      id: 239,
      position: { name: 'Sports', color: 'lime', id: 45 },
      startsAt: '2020-02-05T06:00:00.000Z',
      endsAt: '2020-02-05T13:00:00.000Z',
    },
  ],
  pagination: { offset: 0, limit: 10, count: 9 },
};

export const EventDetails = {
  id: 95,
  startsAt: '2020-06-09T09:00:00.000Z',
  endsAt: '2020-06-09T13:00:00.000Z',
  position: { name: 'Sports', color: 'lime', id: 45 },
  employees: [
    {
      firstName: 'Bettye',
      lastName: 'Mann',
      image: 'https://cdn.fakercloud.com/avatars/vivekprvr_128.jpg',
      id: 73,
    },
  ],
};
