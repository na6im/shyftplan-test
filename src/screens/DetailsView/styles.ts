import styled from 'styled-components';

export const DetailsContainer = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
`;

export const EventDetails = styled.div`
  width: 400px;
  display: flex;
  align-items: flex-start;
  flex-direction: column;
  margin: 10px;
`;

export const EmployeesDetails = styled.div`
  width: 200px;
  display: flex;
  flex-direction: column;
  margin: 10px;
`;

export const EmployeeInfo = styled.div`
  width: 100%;
  display: flex;
  align-items: center;
  flex-direction: row;

  img {
    margin: 10px;
    height: 30px;
    width: 30px;
    border-radius: 30px;
  }
`;

export const NotFound = styled.img`
  height: 100px;
  display: flex;
  align-items: center;
  flex-direction: row;
`;

export const GoBackButton = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: row;
  width: fit-content;
  cursor: pointer;
  border-radius: 8px;
  padding: 5px 10px 5px 5px;
  background-color: rgb(67, 135, 45);
  color: white;
  border: none;
  margin: 10px;

  img {
    height: 15px;
    margin-right: 5px;
  }
`;
