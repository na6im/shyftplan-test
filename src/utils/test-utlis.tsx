// test-utils.js
import React, { ComponentType, ReactChild, ReactChildren } from 'react';
import { render } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from 'react-query';

interface AuxProps {
  children: ReactChild | ReactChildren;
}

const queryClient = new QueryClient();

function renderWithQuery(ui: React.ReactElement) {
  const Wrapper = ({ children }: AuxProps) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );
  return {
    ...render(ui, { wrapper: Wrapper as ComponentType }),
  };
}

// override render method

export * from '@testing-library/react';

export { renderWithQuery };
