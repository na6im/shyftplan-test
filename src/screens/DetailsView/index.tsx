import React from 'react';
import { RouteComponentProps } from 'react-router-dom';
import { useQuery } from 'react-query';
import { AxiosError } from 'axios';
import api from '../../api';
import {
  DetailsContainer,
  EventDetails,
  EmployeesDetails,
  EmployeeInfo,
  NotFound
} from './styles';
import Loader from '../../components/Loader/Loader';
import { LoaderContainer } from '../EventsScreen/styles';
import logo from '../../assets/logo-only.png';
import notFound from '../../assets/not-found.png';
import { EventDetail } from '../../types';

interface MatchParams {
  id: string;
}
const DetailsView: React.FC<RouteComponentProps<MatchParams>> = ({
  match: {
    params: { id },
  },
}) => {
  const { isLoading, data, isError, error } = useQuery<EventDetail, AxiosError>(
    ['events', id],
    () => api.getEvent(id),
    { staleTime: Infinity }
  );

  if (isLoading)
    return (
      <LoaderContainer>
        <Loader src={logo} />
      </LoaderContainer>
    );

  if (isError)
    return (
      <LoaderContainer>
        {error?.response?.status === 404 && (
          <>
            <NotFound src={notFound} alt="not-found" />
            <div>Event not found</div>
          </>
        )}
      </LoaderContainer>
    );

  return (
    <DetailsContainer>
      <EventDetails>
        <div>
          <b>Name: </b> {data?.position?.name}
        </div>
        <div>
          <b>Starts at: </b>
          {new Date(data?.startsAt!).toLocaleString('fr', {
            dateStyle: 'short',
            timeStyle: 'short',
          })}
        </div>
        <div>
          <b>Ends at: </b>
          {new Date(data?.endsAt!).toLocaleString('fr', {
            dateStyle: 'short',
            timeStyle: 'short',
          })}
        </div>
      </EventDetails>
      <EmployeesDetails>
        <b>Employees</b>
        {data?.employees?.map(({ firstName, lastName, image }) => (
          <EmployeeInfo>
            <img src={image} alt={firstName + lastName} /> {firstName}{' '}
            {lastName}
          </EmployeeInfo>
        ))}
      </EmployeesDetails>
    </DetailsContainer>
  );
};

export default DetailsView;
