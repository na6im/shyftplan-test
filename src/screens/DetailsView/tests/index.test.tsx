import React from 'react';
import axios from 'axios';
import { renderWithQuery, cleanup, screen } from '../../../utils/test-utlis';
import DetailsView from '../index';
import { EventDetails } from '../../../__mocks__/mockedData';

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

beforeEach(cleanup);
describe('<DetailsView />', () => {
  /**
   * Unskip this test to use it
   *
   * @see {@link https://jestjs.io/docs/en/api#testskipname-fn}
   */
  it('should show loading then container', async () => {
    mockedAxios.get.mockImplementationOnce(() =>
      Promise.resolve({ data: EventDetails })
    );
    const { getByTestId } = renderWithQuery(
      <DetailsView
        match={{
          params: { id: '126' },
        }}
      />
    );

    expect(getByTestId('isLoading')).toBeInTheDocument();
    expect(await screen.findByTestId(/container/i)).toBeInTheDocument();
  });

  it('should render employees elements', async () => {
    mockedAxios.get.mockImplementationOnce(() =>
      Promise.resolve({ data: EventDetails })
    );
    renderWithQuery(
      <DetailsView
        match={{
          params: { id: '126' },
        }}
      />
    );

    expect(await screen.findByTestId(/container/i)).toBeInTheDocument();
    expect(await screen.findByText(/Employees/i)).toBeInTheDocument();
    const block = await screen.findAllByTestId(/employee/i);
    expect(block).toHaveLength(1);
  });

  it('Should renderWithQuery and match the snapshot', async () => {
    mockedAxios.get.mockImplementationOnce(() =>
      Promise.resolve({
        data: EventDetails,
      })
    );
    const {
      container: { firstChild },
    } = renderWithQuery(
      <DetailsView
        match={{
          params: { id: '126' },
        }}
      />
    );

    expect(await screen.findByTestId(/container/i)).toBeInTheDocument();
    expect(firstChild).toMatchSnapshot();
  });

  it('Should renderWithQuery and match the snapshot', async () => {
    mockedAxios.get.mockImplementationOnce(() =>
      Promise.resolve({
        data: EventDetails,
      })
    );
    const {
      container: { firstChild },
    } = renderWithQuery(
      <DetailsView
        match={{
          params: { id: '126' },
        }}
      />
    );

    expect(await screen.findByTestId(/container/i)).toBeInTheDocument();
    expect(firstChild).toMatchSnapshot();
  });
});
