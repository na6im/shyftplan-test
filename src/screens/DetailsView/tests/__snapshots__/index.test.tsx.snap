// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetailsView /> Should renderWithQuery and match the snapshot 1`] = `
<div
  class="sc-jrsJWt hWrfMW"
  data-testid="isLoading"
>
  <img
    class="sc-gKAaRy bVYUDA"
    src="logo-only.png"
  />
</div>
`;

exports[`<DetailsView /> Should renderWithQuery and match the snapshot 2`] = `
<div
  class="sc-bdnxRM eKZVji"
  data-testid="container"
>
  <button
    class="sc-jSFjdj dgEvJT"
  >
    <img
      alt="back-button"
      src="backIcon.svg"
    />
    Go back
  </button>
  <div
    class="sc-gtsrHT kzWqUm"
  >
    <div>
      <b>
        Name: 
      </b>
       
      Sports
    </div>
    <div>
      <b>
        Starts at: 
      </b>
      09/06/2020 10:00
    </div>
    <div>
      <b>
        Ends at: 
      </b>
      09/06/2020 14:00
    </div>
  </div>
  <div
    class="sc-dlnjwi fpwCtT"
  >
    <b>
      Employees
    </b>
    <div
      class="sc-hKFxyN hoMlYs"
      data-testid="employee"
    >
      <img
        alt="BettyeMann"
        src="https://cdn.fakercloud.com/avatars/vivekprvr_128.jpg"
      />
       
      Bettye
       
      Mann
    </div>
  </div>
</div>
`;
